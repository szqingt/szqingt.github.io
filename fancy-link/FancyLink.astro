---
import { previewFileName } from './utils'
export interface Props {
	href: string
  fancyLinkPaht?: string
  position?: 'up' | 'down'
}

const { href, fancyLinkPaht = '/__fancyLinkPreview', position = 'up' } = Astro.props
const isDevMode = import.meta.env.MODE === 'development'
const previewPath = `${fancyLinkPaht}/${previewFileName(href)}.png`
const darkPreviewPath = `${fancyLinkPaht}/${previewFileName(href, true)}.png`
---
<style>
  .fancy-link-wrap {
    position: relative;
  }
 .fancy-link-wrap .fancy-link:hover +.popover-container{
  display: block;
 }
 .fancy-link-wrap .popover-container img {
  margin: 0;
 }
 .popover-container {
  transform: translateX(50%);
  background-color: var(--c-bg);
 }
 .popover-container.up {
  bottom: 1.5rem;
 }
 .popover-container.down {
  top: 1.5rem;
 }
 :global(.light) .light-preview {
  display: block;
 }
 :global(.dark) .dark-preview {
  display: block;
 }
</style>
<span class="fancy-link-wrap">
  <a class="fancy-link" data-is-fancy-link href={href} target="_blank"><slot /></a>
  <div class:list={[position]} class="popover-container hidden rounded-2 text-center p-2 absolute right-50% w100 h62">
    {isDevMode && <span>Fancy Link Only work in build!</span>}
    {!isDevMode && <img loading="lazy" class="light-preview hidden rounded-2" src={previewPath} alt="link preview"> }
    {!isDevMode && <img loading="lazy" class="dark-preview hidden rounded-2" src={darkPreviewPath} alt="link preview"> }
  </div>
</span>