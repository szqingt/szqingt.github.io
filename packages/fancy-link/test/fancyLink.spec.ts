import { describe, it, expect, afterEach } from 'vitest'
import { resolve } from 'node:path'
import { existsSync, rmSync, readFileSync } from 'node:fs'
import FancyLink, { getPageAllValidFancyLinkInfo } from '../src/fancyLink'
import { readdirSync } from 'node:fs'

const TEST_DATA = [
  {
    attribs: {
      'data-is-fancy-link': 'true',
      href: 'https://google.com',
    },
    href: 'https://google.com',
    name: 'google',
  },
  {
    href: 'https://bing.com',
    name: 'bing',
  },
  {
    href: 'https://google.com',
    name: 'google1',
  },
]

describe('getHtmlPageAllValidUrlForATage', () => {
  it('getHtmlPageAllValidUrlForATage should collect page all Url', () => {
    const allUrl = getPageAllValidFancyLinkInfo(resolve(__dirname, './index.html'))
    expect(allUrl).toMatchSnapshot()
  })
})

// describe('fancy link', () => {

//   // it('fancy link should generate preview file', () => {

//   //   const collect = FancyLink()
//   //   const doneHook = collect.hooks['astro:build:done']
//   //   doneHook && doneHook({
//   //     dir: new URL('fiel:///Users/never_1992/project/szqingt.github.io/fancy-link'),
//   //     pages: [{
//   //       pathname: ''
//   //     }],
//   //     routes: []
//   //   })
//   // })

//   // it('fancy link should generate preview file with target path', () => {

//   //   const collect = FancyLink({
//   //     outPath: resolve(__dirname),
//   //   })
//   //   const doneHook = collect.hooks['astro:build:done']
//   //   doneHook && doneHook({
//   //     dir: new URL('fiel:///Users/never_1992/project/szqingt.github.io/fancy-link'),
//   //     pages: [{
//   //       pathname: ''
//   //     }],
//   //     routes: []
//   //   })
//   // })

//   it('fancy link should generate without google',async () => {

//     const collect = FancyLink({
//       exclude: ['google']
//     })
//     const doneHook = collect.hooks['astro:build:done']
//     doneHook && await doneHook({
//       dir: new URL('fiel:///Users/never_1992/project/szqingt.github.io/fancy-link'),
//       pages: [{
//         pathname: ''
//       }],
//       routes: []
//     })

//     expect(readdirSync(resolve(__dirname, '__fancyLinkPrview'))).toStrictEqual([ '-78bd255e.png' ])
//   }, {
//     timeout: 2 * 60 * 1000
//   })
// })