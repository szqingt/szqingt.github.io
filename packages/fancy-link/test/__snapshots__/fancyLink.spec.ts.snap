// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getHtmlPageAllValidUrlForATage > getHtmlPageAllValidUrlForATage should collect page all Url 1`] = `
[
  {
    "attribs": {
      "href": "https://google.com",
      "mode": "dark",
    },
    "href": "https://google.com",
  },
  {
    "attribs": {
      "href": "https://bing.com",
    },
    "href": "https://bing.com",
  },
  {
    "attribs": {
      "href": "https://google.com",
    },
    "href": "https://google.com",
  },
]
`;
